src/main/java/com/google/cloud/spanner/pgadapter/statements/IntermediateStatement.java

```diff
@@ -240,85 +285,195 @@ public Exception getException() {
-        this.hasMoreData = false;
-        this.statementResult = null;
+        this.hasMoreData[index] = false;
         break;
       default:
         throw SpannerExceptionFactory.newSpannerException(
             ErrorCode.INTERNAL, "Unknown or unsupported result type: " + result.getResultType());
     }
   }
 
-  protected void updateBatchResultCount(long[] updateCounts) {
-    this.updateCount = 0L;
-    for (int i = 0; i < updateCounts.length; ++i) {
-      this.updateCount += updateCounts[i];
+  protected void updateBatchResultCount(int fromIndex, long[] updateCounts) {
+    for (int index = fromIndex; index < fromIndex + updateCounts.length; index++) {
+      setUpdateCount(index, updateCounts[index - fromIndex]);
     }
-    this.hasMoreData = false;
-    this.statementResult = null;
   }
 
   /**
    * Clean up and save metadata when an exception occurs.
    *
    * @param e The exception to store.
    */
-  protected void handleExecutionException(SpannerException e) {
-    this.exception = e;
-    this.hasMoreData = false;
-    this.statementResult = null;
+  protected void handleExecutionException(int index, SpannerException e) {
+    this.exceptions[index] = e;
+    this.hasMoreData[index] = false;
+  }
+
+  private void handleExecutionExceptionInTransaction(int index, SpannerException e) {
+    if (executionStatus == ExecutionStatus.IMPLICIT_TRANSACTION) {
+      connection.rollback();
+      executionStatus = ExecutionStatus.HALTED;
+    } else if (executionStatus == ExecutionStatus.EXPLICIT_TRANSACTION) {
+      connectionHandler.setStatus(ConnectionStatus.ABORTED);
+      executionStatus = ExecutionStatus.HALTED;
+    }
+    handleExecutionException(index, e);
   }
 
   /** Execute the SQL statement, storing metadata. */
   public void execute() {
-    this.executed = true;
-    try {
-      if (statements.size() > 1) {
-        // TODO: Clean this up a little once the statement parsing in the client library has been
-        // released: https://github.com/googleapis/java-spanner/pull/1690
-        // Also, the restriction that mixed batches are not allowed will be removed in a future PR.
-        if (resultType == ResultType.UPDATE_COUNT) {
-          connection.startBatchDml();
-        } else if (resultType == ResultType.NO_RESULT) {
-          connection.startBatchDdl();
-        } else {
-          throw SpannerExceptionFactory.newSpannerException(
-              ErrorCode.INVALID_ARGUMENT, "Statement type is not supported for batching");
-        }
-        for (String stmt : statements) {
-          connection.execute(Statement.of(stmt));
+    if (connectionHandler.getStatus() == ConnectionStatus.ABORTED
+        && !"ROLLBACK".equals(getCommand(0))) {
+      lastExecutedIndex = 0;
+      handleExecutionException(
+          0,
+          SpannerExceptionFactory.newSpannerException(
+              ErrorCode.INVALID_ARGUMENT,
+              "Transaction is aborted and must be rolled back prior to further execution."));
```

Comment:
This is the literal error message that is returned by PostgreSQL

```suggestion
              "ERROR: current transaction is aborted, commands ignored until end of transaction block"));
```

Reply:
Done.